code/game/machinery/machine_frame.dm

```diff
@@ -0,0 +1,472 @@
+	if(!length(replacer.contents))
+		return FALSE
+	var/amt = 0
+	for(var/path in req_components)
+		amt += req_components[path]
+	if(!amt)
+		return FALSE
+
+	var/play_sound = FALSE
+	var/list/part_list = replacer.get_sorted_parts() //parts sorted in order of tier
+	for(var/path in req_components)
+		var/target_path
+		if(ispath(path, /datum/stock_part))
+			var/datum/stock_part/datum_part = path
+			target_path = initial(datum_part.physical_object_base_type)
+		else
+			target_path = path
+
+		var/obj/item/part
+		while(req_components[path] > 0 && (part = look_for(part_list, target_path, ispath(path, /obj/item/stack/ore/bluespace_crystal) ? /obj/item/stack/sheet/bluespace_crystal : null)))
+			part_list -= part
+			if(istype(part, /obj/item/stack))
+				var/obj/item/stack/S = part
+				var/used_amt = min(round(S.get_amount()), req_components[path])
+				var/stack_name = S.singular_name
+				if(!used_amt || !S.use(used_amt))
+					continue
+				req_components[path] -= used_amt
+				// No balloon alert here so they can look back and see what they added
+				to_chat(user, span_notice("You add [used_amt] [stack_name] to [src]."))
+				play_sound = TRUE
+			else if(replacer.atom_storage.attempt_remove(part, src))
+				var/stock_part_datum = GLOB.stock_part_datums_per_object[part.type]
+				if (!isnull(stock_part_datum))
+					components += stock_part_datum
+					qdel(part)
+				else
+					components += part
+					part.forceMove(src)
+				req_components[path]--
+				// No balloon alert here so they can look back and see what they added
+				to_chat(user, span_notice("You add [part] to [src]."))
+				play_sound = TRUE
+
+	if(play_sound && !no_sound)
+		replacer.play_rped_sound()
+		if(replacer.works_from_distance)
+			user.Beam(src, icon_state = "rped_upgrade", time = 0.5 SECONDS)
+	return TRUE
+
+/obj/structure/frame/machine/can_be_unfasten_wrench(mob/user, silent)
+	. = ..()
+	if(. != SUCCESSFUL_UNFASTEN)
+		return .
+
+	if(circuit?.needs_anchored)
+		balloon_alert(user, "frame must be anchored!")
+		return FAILED_UNFASTEN
+
+	return
+
+/obj/structure/frame/machine/screwdriver_act(mob/living/user, obj/item/tool)
+	. = ..()
+	if(. & ITEM_INTERACT_ANY_BLOCKER)
+		return .
+	if(state != FRAME_STATE_BOARD_INSTALLED)
+		return .
+
+	if(finalize_construction(user, tool))
+		return ITEM_INTERACT_SUCCESS
+
+	return ITEM_INTERACT_BLOCKING
+
+/obj/structure/frame/machine/wirecutter_act(mob/living/user, obj/item/tool)
+	. = ITEM_INTERACT_BLOCKING
+	if(state != FRAME_STATE_WIRED)
+		return NONE
+
+	balloon_alert(user, "removing cables...")
+	if(!tool.use_tool(src, user, 2 SECONDS, volume = 50) || state != FRAME_STATE_WIRED)
+		return
+
+	state = FRAME_STATE_EMPTY
+	update_appearance(UPDATE_ICON_STATE)
+	new /obj/item/stack/cable_coil(drop_location(), 5)
+	return ITEM_INTERACT_SUCCESS
+
+/obj/structure/frame/machine/crowbar_act(mob/living/user, obj/item/tool)
+	. = ITEM_INTERACT_BLOCKING
```

Comment:
```suggestion
```
And same here (though this is redundant anyways? Unless it's for runtime protection? If so, should be commented)